src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrNodeResource.java

```diff
@@ -181,34 +185,15 @@ private InputStream getInputStream() {
         final Node node = getNode();
         if (node != null) {
             try {
-                // find the content node: for nt:file it is jcr:content
-                // otherwise it is the node of this resource
-                Node content = (node.isNodeType(NT_FILE) ||
-                                (node.isNodeType(NT_FROZEN_NODE) &&
-                                 node.getProperty(JCR_FROZEN_PRIMARY_TYPE).getString().equals(NT_FILE)))
-                        ? node.getNode(JCR_CONTENT)
-                        : node.isNodeType(NT_LINKED_FILE) ? node.getProperty(JCR_CONTENT).getNode() : node;
-
                 Property data;
-
-                // if the node has a jcr:data property, use that property
-                if (content.hasProperty(JCR_DATA)) {
-                    data = content.getProperty(JCR_DATA);
-                } else {
-                    // otherwise try to follow default item trail
-                    try {
-                        Item item = content.getPrimaryItem();
-                        while (item.isNode()) {
-                            item = ((Node) item).getPrimaryItem();
-                        }
-                        data = (Property) item;
-
-                    } catch (ItemNotFoundException infe) {
-                        // we don't actually care, but log for completeness
-                        LOGGER.debug("getInputStream: No primary items for {}", toString(), infe);
-                        data = null;
-                    }
+                try {
+                    data = getPrimaryProperty(node);
+                } catch (ItemNotFoundException infe) {
+                    // we don't actually care, but log for completeness
+                    LOGGER.debug("getInputStream: No primary items for {}", toString(), infe);
+                    data = null;
```

Comment:
This makes sure that `data` can be passed as null to all URIProvider implementations. 

Does the URIProvider API allow nulls in data. I suspect that other implementations will throw NPE after this change ?

Reply:
This hasn't changed in this PR, even before `data` could have been null (see removed line 209)

Reply:
looking at it again, I think you are right.
The logic doesn't really make sense. Why does it spend time getting `data` when its only needed if convertToPublicURI returns null.

If there are no URIProviders that operation is a noop, but if there are URIProviders then they should be returning a uri.

No change requested here any more.

Reply:
> Why does it spend time getting data when its only needed if convertToPublicURI returns null.

data is needed or both cases!